sbin/st/st.1
.\" Copyright (c) 1981, 1990 The Regents of the University of California.
.\" All rights reserved.
.\"
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)mt.1	6.7 (Berkeley) 7/28/91
.\"
.Dd July 28, 1991
.Dt ST 1
.Os BSD 4
.Sh NAME
.Nm st
.Nd magnetic tape manipulating program
.Sh SYNOPSIS
.Nm st
.Op Fl f Ar tapename
.Ar command
.Op Ar count
.Sh DESCRIPTION
.Nm St
is used to give commands to a magnetic tape drive.
By default
.Nm st
performs the requested operation once.  Operations
may be performed multiple times by specifying
.Ar count  .
Note
that
.Ar tapename
must reference a raw (not block) tape device.
.Pp
The tape drive will hold density and blocking parameters until the next 
unmount. In the case of rst0, this is immediately and is therefore of
not much use with this device. In the case of nrst0 this
may be after several commands when either an offline command is issued or
rst0 is openned and closed. If parameters are to be held across 
unmounts, then they should be made to the control device for each
mode.. i.e. the devices with a submode of 3 (minors 3,7,11,15 for example).
It is suggested that the operator keep the control device unreadble
and unwritable to normal system users, while giving them access to
the non-rewind device (e.g. nrst0) on demand.
.Pp
The available commands are listed below.  Only as many
characters as are required to uniquely identify a command
need be specified.
.Bl -tag -width "eof, weof"
.It Cm eof , weof
Write
.Ar count
end-of-file marks at the current position on the tape.
.It Cm fsf
Forward space
.Ar count
files.
.It Cm fsr
Forward space
.Ar count
records.
.It Cm bsf
Back space
.Ar count
files.
.It Cm bsr
Back space
.Ar count
records.
.It Cm rewind
Rewind the tape
(Count is ignored).
.It Cm offline , rewoffl
Rewind the tape and place the tape unit off-line (possibly eject)
(Count is ignored).
.It Cm blocksize
Sets the block size characteristics of the openned 
device to value specified in
.Ar count.
A 0 means variable sized blocks, and anything else means fixed block, 
with blocksize as that of
.Ar count.
.It Cm density
Set the density for the openned device  (bits 2,3 of minor number)
.Ar count
density code as specified by the SCSI II specification. Valid values
are from 0  to 0x17.
.It The different density codes are as follows:
0x0	default for device
.br
0xE	reserved for ECMA
.br

Value Tracks Density(bpi) Code Type  Reference     Note
.br
0x1     9       800       NRZI  R    X3.22-1983    2
.br
0x2     9      1600       PE    R    X3.39-1986    2
.br
0x3     9      6250       GCR   R    X3.54-1986    2
.br
0x5    4/9     8000       GCR   C    X3.136-1986   1
.br
0x6     9      3200       PE    R    X3.157-1987   2
.br
0x7     4      6400       IMFM  C    X3.116-1986   1
.br
0x8     4      8000       GCR   CS   X3.158-1986   1
.br
0x9    18     37871       GCR   C    X3B5/87-099   2
.br
0xA    22      6667       MFM   C    X3B5/86-199   1
.br
0xB     4      1600       PE    C    X3.56-1986    1
.br
0xC    24     12690       GCR   C    HI-TC1        1,5
.br
0xD    24     25380       GCR   C    HI-TC2        1,5
.br
0xF    15     10000       GCR   C    QIC-120       1,5
.br
0x10   18     10000       GCR   C    QIC-150       1,5
.br
0x11   26     16000       GCR   C    QIC-320(525?) 1,5
.br
0x12   30     51667       RLL   C    QIC-1350      1,5
.br
0x13    1     61000       DDS   CS    X3B5/88-185A 4
.br
0x14    1     43245       RLL   CS    X3.202-1991  4
.br
0x15    1     45434       RLL   CS    ECMA TC17    4
.br
0x16   48     10000       MFM   C     X3.193-1990  1
.br
0x17   48     42500       MFM   C     X3B5/91-174  1
.br

where Code means:
.br
NRZI	Non Return to Zero, change on ones
.br
GCR	Group Code Recording
.br
PE	Phase Encoded
.br
IMFM	Inverted Modified Frequency Modulation
.br
MFM	Modified Frequency Modulation
.br
DDS	Dat Data Storage
.br
RLL	Run Length Encoding
.br

where Type means:
.br
R	Real-to-Real
.br
C	Cartridge
.br
CS	cassette
.br

where Notes means:
.br
1	Serial Recorded
.br
2	Parallel Recorded
.br
3	Old format know as QIC-11
.br
4	Helical Scan
.br
5	Not ANSI standard, rather industry standard.
.br

.It Cm status
Print status information about the tape unit.
Information is printed out about the present active parameters and
also the four Operating modes available.
.El
.Pp
If a tape name is not specified, and the environment variable
.Ev TAPE
does not exist;
.Nm st
uses the device
.Pa /dev/nrst0 .
.Pp
.Nm St
returns a 0 exit status when the operation(s) were successful,
1 if the command was unrecognized, and 2 if an operation failed.
.Sh ENVIRONMENT
If the following environment variable exists, it is utilized by
.Nm st .
.Bl -tag -width Fl
.It Ev TAPE
.Nm St
checks the
.Ev TAPE
environment variable if the
argument
.Ar tapename
is not given.
.Sh FILES
.Bl -tag -width /dev/rmt* -compact
.It Pa /dev/rst*
Raw magnetic tape interface
.El
.Sh SEE ALSO
.\".Xr mtio 4 ,
.Xr st 4 ,
.Xr dd 1 ,
.Xr ioctl 2 ,
.Xr environ 7
.Sh HISTORY
The
.Nm st
command appeared in 386BSD 0.1.

.\" mt.1: mtio(4) missing


NEWS
# Major changes between releases

## Changes in version 0.15.0

**Released on 2026/10/07.**

### Bugfixes

* #265: Null cleanups
* #271: kyua: Try harder to delete directories.
* #300: Implement utils::config::positive_int_node::deep_copy

### Developer Workflow Changes

* #254: Add initial GitHub Actions support for kyua
* #288: [CI] remove legacy CI scripts and files
* #291: Remove Travis CI config
* #298: Add sanitizers and code coverage support
* #301: Add FreeBSD CI support via vmactions
* #305: Automate uploading release assets

### Feature Enhancements

* #203: Add GoogleTest support
* #230: Do not count skipped as passed in test cmd
* #260: Support require.diskspace ATF metadata
* #261: Pass unprivileged user config prop to ATF using all known names
* #266: Add "kyua debug -p" option
* #270: Add required_kmods metadata
* #272: Make "test" command actually report about workdir cleanup issues
* #273: Make "debug -p" skip writing stdout/stderr to tmp files
* #277: kyua-debug: Add -P option
* #290: Require C++-20 in order to build kyua
* #307: Upstream multiple items from FreeBSD

### Documentation

* #258: kyuafile.5: Add ATF metadata mapping reference
* #275: kyua manuals: Correct --results-file short form
* #279: docs: Standardize document description tone

### Miscellaneous

* #264: Switch from std::auto_ptr<> to std::unique_ptr<>
* #269: Fix cmd_about_test:topic__version test case
* #297: Reduce overall lutok dependency
* #312: Replace KYUA_ATTRIBUTE_{NORETURN,UNUSED}} with C++ attributes
* #318: utils::fs: remove umount(8) support

**Full Changelog**: https://github.com/freebsd/lutok/compare/kyua-0.14.1...kyua-0.15.0

## Changes in version 0.14.1

**Released on March 29th, 2025.**

* Issue #263: fix release tarball by relying on `make dist`.

* Document ATF to Kyua mappings for metadata properties.

* Support `require.diskspace` ATF metadata property.

* Support backwards compatible `unprivileged_user` metadata property for ease
  of use dealing with ATF vs Kyua.

* Fix building from source by tracking `utils/process/Kyuafile.in` and
  `engine/execenv/execenv_fwd.hpp` with automake.

## Changes in version 0.14

**Released on December 31st, 2024.**

* Replace `--with-atf` with `--enable-atf`. The latter must be specified
  to enable ATF support and the Kyua tests by proxy.

* Require newer versions of ATF (0.21) and Lutok (0.5). This is being done
  as part of a multi-phase deprecation of pre-C++11 environments.

* Vendor the ATF m4 files from the freebsd/atf repo.

* Translate `is_exclusive` metadata to its ATF counterpart.

* Issue #236: fix accidental test forkbomb in specific user environments
  with `integration` tests by avoiding infinite recursion calling kyua.

* Fix the build on Ubuntu 24.04.

- execenv support added. This FreeBSD-specific feature allows tests to be
  executed in jails.

* Issue #169: Remove deprecated `readdir_r` use.

* Fix build issue on more recent versions of MacOS.

* Fix build issue on IllumOS.

* Fix issues with PID reuse in testcases.

* Escape characters captured from stderr and stdout in HTML output.

* `integration/cmd_report_test:verbose`: fix regular expression used in
  tests which would break in certain user environments.

* Add Cirrus CI support for testing Kyua on FreeBSD.

* Remove `UTILS_UNUSED_PARAM` macro.

* Fix document rendering with newer versions of doxygen.

* Issue #152: explicitly require C++11 language features when compiling Kyua.
  Remove `std::shared_ptr` compatibility header, `utils/shared_ptr.hpp`.

* Fix `igor` and `mandoc -Tlint` reported issues with generated manpages.

* Issue #167: fix typos in kyua(1) manpage.

## Changes in version 0.13

**Released on August 26th, 2016.**

* Fixed execution of test cases as an unprivileged user, at least under
  NetBSD 7.0.  Kyua-level failures were probably a regression introduced
  in Kyua 0.12, but the underlying may have existed for much longer:
  test cases might have previously failed for mysterious reasons when
  running under an unprivileged user.

* Issue #134: Fixed metadata test broken on 32-bit platforms.

* Issue #139: Added per-test case start/end timestamps to all reports.

* Issue #156: Fixed crashes due to the invalid handling of cleanup
  routine data and triggered by the reuse of PIDs in long-running Kyua
  instances.

* Issue #159: Fixed TAP parser to ignore case while matching `TODO` and
  `SKIP` directives, and to also recognize `Skipped`.

* Fixed potential crash due to a race condition in the unprogramming of
  timers to control test deadlines.


## Changes in version 0.12

**Released on November 22nd, 2015.**

This is a huge release and marks a major milestone for Kyua as it finally
implements a long-standing feature request: the ability to execute test
cases in parallel.  This is a big deal because test cases are rarely
CPU-bound: running them in parallel yields much faster execution times for
large test suites, allowing faster iteration of changes during development.

As an example: the FreeBSD test suite as of this date contains 3285 test
cases.  With sequential execution, a full test suite run takes around 12
minutes to complete, whereas on a 4-core machine with a high level of
parallelism it takes a little over 1 minute.

Implementing parallel execution required rewriting most of Kyua's core and
partly explains explains why there has not been a new release for over a
year.  The current implementation is purely subprocess-based, which works
but has some limitations and has resulted in a core that is really complex
and difficult to understand.  Future versions will investigate the use of
threads instead for a simplified programming model and additional
parallelization possibilities.

* Issue #2: Implemented support to execute test cases in parallel when
  invoking `kyua test`.  Parallel execution is *only* enabled when the new
  `parallelism` configuration variable is set to a value greater than `1`.

sbin/st/st.c
/*
 * Copyright (c) 1980 The Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1980 The Regents of the University of California.\n\
 All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)mt.c	5.6 (Berkeley) 6/6/91";
#endif /* not lint */

/*
 * mt --
 *   magnetic tape manipulation program
 */
#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/mtio.h>
#include <fcntl.h>
#include <stdio.h>
#include <ctype.h>

#define	equal(s1,s2)	(strcmp(s1, s2) == 0)

struct commands {
	char *c_name;
	int c_code;
	int c_ronly;
} com[] = {
	{ "weof",	MTWEOF,	0 },
	{ "eof",	MTWEOF,	0 },
	{ "fsf",	MTFSF,	1 },
	{ "bsf",	MTBSF,	1 },
	{ "fsr",	MTFSR,	1 },
	{ "bsr",	MTBSR,	1 },
	{ "rewind",	MTREW,	1 },
	{ "offline",	MTOFFL,	1 },
	{ "rewoffl",	MTOFFL,	1 },
	{ "status",	MTNOP,	1 },
	{ "blocksize",	MTSETBSIZ,	0 },
	{ "density",	MTSETDNSTY,	0 },
	{ 0 }
};

int mtfd;
struct mtop mt_com;
struct mtget mt_status;
char *tape;

main(argc, argv)
	char **argv;
{
	void usage();
	char line[80], *getenv();
	register char *cp;
	register struct commands *comp;

	if (argc > 2 && (equal(argv[1], "-t") || equal(argv[1], "-f"))) {
		argc -= 2;
		tape = argv[2];
		argv += 2;
	} else
		if ((tape = getenv("TAPE")) == NULL)
			tape = DEFTAPE;
	if (argc < 2) {
		usage();
		exit(1);
	}
	cp = argv[1];
	if ((strncmp(cp, "blocksize", strlen(cp)) == 0) && argc < 3 ) {
		usage();
		exit(1);
	}
	if ((strncmp(cp, "density", strlen(cp)) == 0) && argc < 3 ) {
		usage();
		exit(1);
	}
	for (comp = com; comp->c_name != NULL; comp++)
		if (strncmp(cp, comp->c_name, strlen(cp)) == 0)
			break;
	if (comp->c_name == NULL) {
		fprintf(stderr, "mt: don't grok \"%s\"\n", cp);
		usage();
		exit(1);
	}
	if ((mtfd = open(tape, comp->c_ronly ? O_RDONLY : O_RDWR)) < 0) {
		perror(tape);
		exit(1);
	}
	if (comp->c_code != MTNOP) {
		mt_com.mt_op = comp->c_code;
		mt_com.mt_count = (argc > 2 ? atoi(argv[2]) : 1);
		if (mt_com.mt_count < 0) {
			fprintf(stderr, "mt: negative repeat count\n");
			exit(1);
		}
		if (ioctl(mtfd, MTIOCTOP, &mt_com) < 0) {
			fprintf(stderr, "%s %s %d ", tape, comp->c_name,
				mt_com.mt_count);
			perror("failed");
			exit(2);
		}
	} else {
		if (ioctl(mtfd, MTIOCGET, (char *)&mt_status) < 0) {
			perror("mt");
			exit(2);
		}
		status(&mt_status);
	}
}

#ifdef vax
#include <vaxmba/mtreg.h>
#include <vaxmba/htreg.h>

#include <vaxuba/utreg.h>
#include <vaxuba/tmreg.h>
#undef b_repcnt		/* argh */
#include <vaxuba/tsreg.h>
#endif

#ifdef sun
#include <sundev/tmreg.h>
#include <sundev/arreg.h>
#endif

#ifdef tahoe
#include <tahoe/vba/cyreg.h>
#endif

struct tape_desc {
	short	t_type;		/* type of magtape device */
	char	*t_name;	/* printing name */
	char	*t_dsbits;	/* "drive status" register */
	char	*t_erbits;	/* "error" register */
} tapes[] = {
#ifdef vax
	{ MT_ISTS,	"ts11",		0,		TSXS0_BITS },
	{ MT_ISHT,	"tm03",		HTDS_BITS,	HTER_BITS },
	{ MT_ISTM,	"tm11",		0,		TMER_BITS },
	{ MT_ISMT,	"tu78",		MTDS_BITS,	0 },
	{ MT_ISUT,	"tu45",		UTDS_BITS,	UTER_BITS },
#endif
#if defined(sun) 
	{ MT_ISCPC,	"TapeMaster",	TMS_BITS,	0 },
	{ MT_ISAR,	"Archive",	ARCH_CTRL_BITS,	ARCH_BITS },
#endif
#ifdef tahoe
	{ MT_ISCY,	"cipher",	CYS_BITS,	CYCW_BITS },
#endif
#if  defined (__386BSD__) 
	{ MT_ISAR,	"Archive/Tandberg?",	0,	0 },
#endif
	{ 0 }
};

/*
 * Interpret the status buffer returned
 */
status(bp)
	register struct mtget *bp;
{
	register struct tape_desc *mt;

	for (mt = tapes; mt->t_type; mt++)
		if (mt->t_type == bp->mt_type)
			break;
	if (mt->t_type == 0) {
		printf("unknown tape drive type (%d)\n", bp->mt_type);
		return;
	}
	printf("%s tape drive, residual=%d, blocksize=%d\n", 
		mt->t_name, bp->mt_resid, bp->mt_bsiz);
	printf("Density: dflt(0) = %d(0x%x) 1 = %d(0x%x), 2 = %d(0x%x), 3 = %d(0x%x)\n", 
		bp->mt_dns_dflt, bp->mt_dns_dflt,
		bp->mt_dns_dsty1, bp->mt_dns_dsty1,
		bp->mt_dns_dsty2, bp->mt_dns_dsty2,
		bp->mt_dns_dsty3, bp->mt_dns_dsty3);
	printreg("ds", bp->mt_dsreg, mt->t_dsbits);
	printreg("\ner", bp->mt_erreg, mt->t_erbits);
	putchar('\n');
}

/*
 * Print a register a la the %b format of the kernel's printf
 */
printreg(s, v, bits)
	char *s;
	register char *bits;
	register unsigned short v;
{
	register int i, any = 0;
	register char c;

	if (bits && *bits == 8)
		printf("%s=%o", s, v);
	else
		printf("%s=%x", s, v);
	bits++;
	if (v && bits) {
		putchar('<');
		while (i = *bits++) {
			if (v & (1 << (i-1))) {
				if (any)
					putchar(',');
				any = 1;
				for (; (c = *bits) > 32; bits++)
					putchar(c);
			} else
				for (; *bits > 32; bits++)
					;
		}
		putchar('>');
	}
}
void usage() {
	register struct commands *comp;

	fprintf(stderr, "Usage: st [ -f tape ] command [ count ] \n");
	fprintf(stderr, "	Where command is one of:\n");
	for (comp=com; comp->c_name != NULL; comp++) {
		fprintf(stderr, "	%s\n", comp->c_name);
	}
	fprintf(stderr,"Note that the count argument is required\n");
	fprintf(stderr, "with the \"blocksize\" , and the density setting commands.\n");
	fprintf(stderr, "Note that the count argument is a base 10 number\n");
}
